src/herder/TxSetFrame.cpp

```diff
@@ -136,23 +137,51 @@ TxSetFrame::sortForApply()
     return retList;
 }
 
-struct SurgeSorter
+struct AccountTransactionQueue
 {
-    map<AccountID, double>& mAccountFeeMap;
-    SurgeSorter(map<AccountID, double>& afm) : mAccountFeeMap(afm)
+    AccountID mID;
+    std::deque<TransactionFramePtr> mTxs;
+};
+
+struct SurgeCompare
+{
+    Hash mSeed;
+    LedgerTxnHeader const& mHeader;
+    SurgeCompare(Hash const& seed, LedgerTxnHeader const& header)
+        : mSeed(seed), mHeader(header)
     {
     }
 
+    // return true if tx1 < tx2
     bool
-    operator()(TransactionFramePtr const& tx1, TransactionFramePtr const& tx2)
+    operator()(AccountTransactionQueue const* tx1,
+               AccountTransactionQueue const* tx2)
     {
-        if (tx1->getSourceID() == tx2->getSourceID())
-            return tx1->getSeqNum() < tx2->getSeqNum();
-        double fee1 = mAccountFeeMap[tx1->getSourceID()];
-        double fee2 = mAccountFeeMap[tx2->getSourceID()];
-        if (fee1 == fee2)
-            return tx1->getSourceID() < tx2->getSourceID();
-        return fee1 > fee2;
+        if (tx1 == nullptr || tx1->mTxs.empty())
+        {
+            return !tx2->mTxs.empty();
+        }
+        if (tx2 == nullptr || tx2->mTxs.empty())
+        {
+            return false;
+        }
+
+        auto& top1 = tx1->mTxs.front();
+        auto& top2 = tx2->mTxs.front();
+
+        // compare fee/minFee between top1 and top2
+        auto v1 = bigMultiply(top1->getFee(), top2->getMinFee(mHeader));
+        auto v2 = bigMultiply(top2->getFee(), top1->getMinFee(mHeader));
+        if (v1 < v2)
+        {
+            return true;
+        }
+        else if (v1 > v2)
+        {
+            return false;
+        }
+        // use hash of transaction as a tie breaker
+        return lessThanXored(top1->getFullHash(), top2->getFullHash(), mSeed);
```

Comment:
I couldn't entirely follow where the conversation with Jeremy wound up on this but I _think_ using a variant of `shortHash::computeHash` might offer a little more strength.